convex/lp/simplex.go

```diff
@@ -0,0 +1,608 @@
+		for replace, v := range move {
+			if v > blandZeroTol {
+				continue
+			}
+			copy(biCopy, basicIdxs)
+			biCopy[replace] = nonBasicIdx[minIdx]
+			abTmp := extractColumns(A, biCopy)
+			// If the condition number is reasonable, use this index.
+			if mat64.Cond(abTmp, 1) < 1e16 {
+				return replace, minIdx, nil
+			}
+		}
+	}
+	return -1, -1, ErrBland
+}
+
+func verifyInputs(initialBasic []int, c []float64, A mat64.Matrix, b []float64) error {
+	m, n := A.Dims()
+	if len(c) != n {
+		panic("lp: c vector incorrect length")
+	}
+	if len(b) != m {
+		panic("lp: b vector incorrect length")
+	}
+	if len(c) != n {
+		panic("lp: c vector incorrect length")
+	}
+	if len(initialBasic) != 0 && len(initialBasic) != m {
+		panic("lp: initialBasic incorrect length")
+	}
+
+	// Do some sanity checks so that ab does not become singular during the
+	// simplex solution. If the ZeroRow checks are removed then the code for
+	// finding a set of linearly indepent columns must be improved.
+
+	// Check that if a row of A only has zero elements that corresponding
+	// element in b is zero, otherwise the problem is infeasible.
+	// Otherwise return ErrZeroRow.
+	for i := 0; i < m; i++ {
+		isZero := true
+		for j := 0; j < n; j++ {
+			if A.At(i, j) != 0 {
+				isZero = false
+				break
+			}
+		}
+		if isZero && b[i] != 0 {
+			// Infeasible
+			return ErrInfeasible
+		} else if isZero {
+			return ErrZeroRow
+		}
+	}
+	// Check that if a column only has zero elements that the respective C vector
+	// is positive (otherwise unbounded). Otherwise return ErrZeroColumn.
+	for j := 0; j < n; j++ {
+		isZero := true
+		for i := 0; i < m; i++ {
+			if A.At(i, j) != 0 {
+				isZero = false
+				break
+			}
+		}
+		if isZero && c[j] < 0 {
+			return ErrUnbounded
+		} else if isZero {
+			return ErrZeroColumn
+		}
+	}
+	return nil
+}
+
+// initializeFromBasic initializes the basic feasible solution given a set of
+// basic indices. It extracts the columns of A specified by basicIdxs and finds
+// the x values at that location. If the columns of A are not linearly independent
+// or if the initial set is not feasible, valid is false.
+func initializeFromBasic(ab *mat64.Dense, b []float64) (xb []float64, err error) {
+	m, _ := ab.Dims()
+	xb = make([]float64, m)
+	xbMat := mat64.NewVector(m, xb)
+
+	err = xbMat.SolveVec(ab, mat64.NewVector(m, b))
+	if err != nil {
+		return nil, errors.New("lp: subcolumns of A for supplied initial basic singular")
+	}
+	// The solve ensures that the equality constraints are met (ab * xb = b).
+	// Thus, the solution is feasible if and only if all of the x's are positive.
+	allPos := true
+	for _, v := range xb {
+		if v < -initPosTol {
+			allPos = false
+			break
+		}
+	}
+	if !allPos {
+		return xb, errors.New("lp: supplied subcolumns not a feasible solution")
+	}
+	return xb, nil
+}
+
+// extractColumns creates a new matrix out of the columns of A specified by cols.
+func extractColumns(A mat64.Matrix, cols []int) *mat64.Dense {
```

Comment:
It seems to me that this could take the result as a parameter and thus avoid creating garbage at a few places.